### PR TITLE
Support for Fusion copy and move operations

### PR DESCRIPTION
**Pull request by tlemo**
This PR implements copy and move operations for Fusion objects. The intention is to provide a generic container view of Fusion objects, ie. allow them to be copied and moved similarly to std::vector or std::unordered_map containers:

1. Copies are supported, but relatively expensive
2. Move operations are cheap (and noexcept)
3. Fusion::clear() can be used to reset the IR to a "blank state" (also noexcept)
4. Fusion supports swap() as well.

The cloning machinery is implemented as non-intrusive as possible with the help of the new IrCloner class + a new "cloning" constructor in each IR node type (ex. Statement::Statement(const Statement* src, IrCloner* ir_cloner))

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/fusion.cpp [hunk 1]
```diff
@@ -57,14 +59,141 @@ std::unordered_set<Val*> InputsOf::output(Fusion* fusion, Val* output_) {
...
+  // Fixup the Statement::fusion_ links for b
+  for (auto val : b.val_set_) {
+    val->fusion_ = &b;
+  }
+  for (auto expr : b.expr_set_) {
+    expr->fusion_ = &b;
+  }
+}
+
+Fusion::Fusion(const Fusion& other) : IrInputOutput() {
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/fusion.cpp]: What clang-tidy is complaining about is generally reasonable, although not in this case (calling the default IrInputOutput constructor is intentional)

Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/fusion.cpp]: Why do we need to explicitly call default constructor of inherited class? I thought this is the default behavior.

File: torch/csrc/jit/codegen/cuda/ir_base_nodes.h [hunk 2]
```diff
@@ -23,7 +23,7 @@
 /*
  * This file defines the base IR structure. Any IR node in this system will
  * inherit from one of the following classes: Statement, Expr, Val,
- * IRInputOutput IR is any information that the code generation stack may need
+ * IrInputOutput IR is any information that the code generation stack may need
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/ir_base_nodes.h:26]: Looks like `IrInputOutput` should be removed here

Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/ir_base_nodes.h:26]: Looks like IrInputOutput should be removed here

File: torch/csrc/jit/codegen/cuda/fusion.cpp (hunk 1 above)
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/fusion.cpp]: @jjsjann123 : The reason I called the default constructor explicitly is to (1) make it obvious we're intentionally doing something unusual and (2) attempted to suppress compiler warnings (as you can see, I wasn't able to defeat clang-tidy though)

Long story: you're correct, it's the default. For copy/move constructors, this default behavior is error-prone though: in most cases you want to call the base copy (or move) constructor, not the default constructor (and most compilers/linters will happily point this out)

In this case though, we're doing a deep copy, something that doesn't easily subdivide across a class hierarchy (ex. calling IrInputOutput's copy constructor would require to somehow pass the IrCloner to it, which is not really possible). So in the original version I let IrInputOutput construct an empty base object and cloned it explicitly. 

This seems a minor issue, but it was one of the many red flags around IrInputOutput, so I ended up refactoring it out completely (which eliminated a few other problems as well)

File: torch/csrc/jit/codegen/cuda/fusion.cpp [hunk 3]
```diff
@@ -57,6 +59,103 @@ std::unordered_set<Val*> InputsOf::output(Fusion* fusion, Val* output_) {
   return io.inputs;
 }
 
+void swap(Fusion& a, Fusion& b) noexcept {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/fusion.cpp:63]: nice

File: torch/csrc/jit/codegen/cuda/ir_cloner.h [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+  void handle(const Allocate*) override;
+
+  void handle(const Split*) override;
+  void handle(const Merge*) override;
+
+ private:
+  // The destination Fusion container
+  Fusion* fusion_ = nullptr;
+
+  // The dispatch interface doesn't allow returning values from
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/ir_cloner.h:79]: Could you use the mutate dispatch here?

File: torch/csrc/jit/codegen/cuda/ir_cloner.cpp [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+    // since the cloning can be reentrant
+    auto saved_clone = clone_;
+    handle(statement);
+    auto new_node = clone_;
+    clone_ = saved_clone;
+
+    // The base cloning constructor (Statement) should have
+    // registered the new node
+    TORCH_CHECK(new_node != nullptr);
+    TORCH_CHECK(clones_map_[statement] == new_node);
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/ir_cloner.cpp]: Should these be TORCH_INTERNAL_ASSERT? Can you add a comment as to why it might have failed?

File: torch/csrc/jit/codegen/cuda/ir_base_nodes.cpp [hunk 6]
```diff
@@ -19,6 +19,12 @@ namespace torch {
 namespace jit {
 namespace fuser {
 
+Statement::Statement(const Statement* src, IrCloner* ir_cloner) {
+  ir_cloner->registerClone(src, this);
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/ir_base_nodes.cpp:23]: Couldn't registerClone be called inside ir_cloner? Then couldn't we implement `unsafeCloneAs` on all the IR nodes as protected functions? Might just keep IrCloner more isolated from the nodes themselves, although it would require a friend, and I know some people don't like friends.

File: torch/csrc/jit/codegen/cuda/ir_cloner.h [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+class TORCH_CUDA_API IrCloner : private OptInConstDispatch {
+  friend struct Statement;
+
+ public:
+  explicit IrCloner(Fusion* new_fusion) : fusion_(new_fusion) {}
+
+  Statement* clone(const Statement* statement);
+
+  template <class T>
+  T* clone(const T* node) {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/ir_cloner.h:26]: Sorry, having a hard time understanding what this is used for and where.

File: torch/csrc/jit/codegen/cuda/ir_iostream.cpp [hunk 8]
```diff
@@ -8,21 +8,20 @@ namespace torch {
 namespace jit {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/ir_iostream.cpp:8]: good cleanup below, but please isolate in separate PRs.

File: torch/csrc/jit/codegen/cuda/ir_iostream.cpp [hunk 9]
```diff
@@ -44,13 +43,16 @@ void IRPrinter::handle(const Expr* e) {
 void IRPrinter::printHeader(Fusion* fusion, const std::string& kernel_name_) {
   os << "__global__ void " << kernel_name_ << "(";
 
-  std::deque<Val*> vals;
-  for (decltype(fusion->nInputs()) i{0}; i < fusion->nInputs(); i++)
-    vals.push_back(fusion->input(i));
-  for (decltype(fusion->nOutputs()) i{0}; i < fusion->nOutputs(); i++)
-    vals.push_back(fusion->output(i));
+  std::vector<Val*> values;
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/ir_iostream.cpp]: Nit: I typically use vals as it's literally multiple `Val` nodes

File: torch/csrc/jit/codegen/cuda/ir_base_nodes.cpp (hunk 6 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/ir_base_nodes.cpp:23]: Never mind, realized this wouldn't work later on as clone is protecting against duplicate clones and you're calling clone inside the constructors with the IrCloner. Leaving this here just so you can confirm I understood correctly.

File: torch/csrc/jit/codegen/cuda/fusion.cpp [hunk 10]
```diff
@@ -57,6 +59,103 @@ std::unordered_set<Val*> InputsOf::output(Fusion* fusion, Val* output_) {
...
+    auto from_val = ir_cloner.clone(kv.first);
+    auto to_val = ir_cloner.clone(kv.second);
+    values_map_.insert({from_val, to_val});
+  }
+
+  inputs_ = ir_cloner.clone(other.inputs_);
+  outputs_ = ir_cloner.clone(other.outputs_);
+}
+
+Fusion::Fusion(Fusion&& other) noexcept {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/fusion.cpp:143]: Any specific use for move constructors you have in mind, or just wanted to support them?

File: torch/csrc/jit/codegen/cuda/ir_base_nodes.cpp [hunk 11]
```diff
@@ -262,6 +204,25 @@ Expr::Expr(ExprType _type) : type_{_type} {
   this->fusion_ = fusion;
 }
 
+Expr::Expr(const Expr* src, IrCloner* ir_cloner)
+    : Statement(src, ir_cloner),
+      type_(src->type_),
+      inputs_(ir_cloner->clone(src->inputs_)),
+      outputs_(ir_cloner->clone(src->outputs_)) {}
+
+bool Expr::sameAs(const Expr* const other) const {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/ir_base_nodes.cpp:213]: I realize this is my fault, but I don't think this is safe to call. Realizing binaryOp, unaryOp, and ternaryOp won't evaluate correctly here. Would you mind just returning `this==other`? as I think that's the best we can do without implementing dispatch based checking here.

**Review by csarofeen — CHANGES_REQUESTED** [review]
Overall I think this is good, 1 minor thing:
I believe in ir_base_nodes.h there's a section listing everything that needs to be done to add a new node type. Could you add the new required constructor and graph viz?
1 a bit more major thing:
Please don't use the copy constructor, and use an explicit clone on Fusion. I'd like to follow the python semantics of working with our interface objects. This I feel pretty strongly about, so if you disagree it's going to have to be from and end user perspective.

**Comment by csarofeen** [comment]
Just FYI for review I changed the branch this PR is relative to.

**Comment by tlemo** [comment]
> Overall I think this is good, 1 minor thing:
> I believe in ir_base_nodes.h there's a section listing everything that needs to be done to add a new node type. Could you add the new required constructor and graph viz?

Good point, I'll add the notes.

> 1 a bit more major thing:
> Please don't use the copy constructor, and use an explicit clone on Fusion. I'd like to follow the python semantics of working with our interface objects. This I feel pretty strongly about, so if you disagree it's going to have to be from and end user perspective.

Can you please elaborate: what do you mean by python semantics and what is the interface you have in mind?

File: torch/csrc/jit/codegen/cuda/ir_cloner.h (hunk 4 above)
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/ir_cloner.h:79]: Good point. We'd need a new OptInConstMutator, plus change the mutations map to const* Statement -> Statement*

It could be done, but it may not be a net simplification for a single use. What do you think? 

BTW, the whole IrCloner::clone_ is mostly unnecessary at this point, the only use is to double check that we registered the right mapping.

PS. Since we are on topic, why is OptInMutator::mutations mapping Val* and not Statement* ?

File: torch/csrc/jit/codegen/cuda/ir_cloner.cpp (hunk 5 above)
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/ir_cloner.cpp]: I'm not sold on the TORCH_INTERNAL_ASSERT vs TORCH_CHECK distinction, so I prefer to use TORCH_CHECK by default, and change them when explicitly requested. Such in this case - fixed.

These are paranoid checks and I contemplated getting rid of the IrCloner::clone_ completely.

File: torch/csrc/jit/codegen/cuda/ir_base_nodes.cpp (hunk 6 above)
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/ir_base_nodes.cpp:23]: Yes, this goes a bit deeper:

(1) registration needs to be the very first thing done after allocating a Statement subclass, otherwise we wouldn't be able to handle cycles in the IR graph (ex. ForLoop::parent_scope_)

(2) in order to split the individual node cloning outside the nodes themselves you need not only friend access but you also need good default constructors for every nodes. It would also require us to remove all the const data members (ex. ForLoop::iter_domain_) since you can't mutate a const after it's constructed. And I'm not sure if it would be a good idea to put something so intimately tied to the members of a class far away from the class itself.

File: torch/csrc/jit/codegen/cuda/ir_cloner.h (hunk 7 above)
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/ir_cloner.h:26]: Convenience + type safety. IrCloner::clone(const Statement* statement) happy clones any Statement* (including derived classes) and returns a Statement*.

W/o this helper template you'd need to type everywhere:

`val = ir_cloner->clone(src)->as<Val>()`

With it, we can simply say

`val = ir_cloner->clone(src)`

File: torch/csrc/jit/codegen/cuda/ir_iostream.cpp (hunk 8 above)
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/ir_iostream.cpp:8]: This small rewrite was required since we removed IRInputOutput.

File: torch/csrc/jit/codegen/cuda/ir_iostream.cpp (hunk 9 above)
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/ir_iostream.cpp]: Having two local variables named `val` and `vals` seems cruel to human readers.

File: torch/csrc/jit/codegen/cuda/fusion.cpp (hunk 10 above)
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/fusion.cpp:143]: For containers where copies are expensive, move operations can make a significant improvement in the quality of life. You can easily say

`Fusion fusion = createSchedule(base_fusion);`

Instead of being forced to resort to ugly contortions (ex. out parameters).

In general, and I found this to be an important guideline, implementing resource ownership is much easier as an all-or-nothing proposition. This rule even has it's own name: https://en.wikipedia.org/wiki/Rule_of_three_(C%2B%2B_programming)

TL;DR: it may make sense to have move w/o copy, but there's no excuse for copy w/o move.

File: torch/csrc/jit/codegen/cuda/ir_base_nodes.cpp (hunk 11 above)
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/ir_base_nodes.cpp:213]: I'm not sure I follow. There are no new calls to sameAs() in this PR, it's been just moved out of the header (not only to simplify the class definition, but it also allows calling the auto-return functions like Expr::inputs() w/o any concerns regarding the declaration ordering)

File: torch/csrc/jit/codegen/cuda/ir_cloner.h (hunk 4 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/ir_cloner.h:79]: I think we can definitely hold off on this for now and keep it in mind. Not sure on the P.S. as I don't use mutate too often, I think it's used in lowering though we could check if it makes any sense there, otherwise maybe just a mistake.

File: torch/csrc/jit/codegen/cuda/ir_cloner.cpp (hunk 5 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/ir_cloner.cpp]: I would rather err on the side of paranoid checks.

File: torch/csrc/jit/codegen/cuda/ir_base_nodes.cpp (hunk 6 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/ir_base_nodes.cpp:23]: okay, makes sense, thanks.

File: torch/csrc/jit/codegen/cuda/ir_cloner.h (hunk 7 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/ir_cloner.h:26]: ah okay, thanks!

File: torch/csrc/jit/codegen/cuda/ir_iostream.cpp (hunk 9 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/ir_iostream.cpp]: 🤷‍♂️ maybe I'm just cruel then, but when having multiples of a thing I like to just append the s to the thing

File: torch/csrc/jit/codegen/cuda/ir_base_nodes.cpp (hunk 11 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/ir_base_nodes.cpp:213]: This function is bad and my fault. Was just wondering since you touched it if you would simplify it to something that would actually be more correct i.e. `return this==other;`

**Comment by csarofeen** [comment]
Almost all objects in PyTorch even in the c++ backend follow this kind of pattern:
```
  auto options = at::TensorOptions().dtype(at::kFloat);

  at::Tensor t0 = at::ones({2, 2}, options);
  std::cout<<t0<<std::endl;
  auto t1 = t0;
  t1[0][0]=-2;
  std::cout<<t0<<std::endl;
  auto t2 = t1.clone();
  t2[0][1] = -3;
  std::cout<<t0<<std::endl;
  std::cout<<t2<<std::endl;
```
Output:
```
 1  1
 1  1
[ CPUFloatType{2,2} ]
-2  1
 1  1
[ CPUFloatType{2,2} ]
-2  1
 1  1
[ CPUFloatType{2,2} ]
-2 -3
 1  1
[ CPUFloatType{2,2} ]
```

Copies are only reference copies like they would be in ~pytorch~ python, not deep copies unless explicitly using clone().

File: test/cpp/jit/test_gpu.cpp [hunk 12]
```diff
@@ -426,6 +426,136 @@ void testGPU_FusionClear() {
...
+
+    tv3->axis(0)->parallelize(ParallelType::BIDx);
+    tv3->axis(-1)->parallelize(ParallelType::TIDx);
+  }
+
+  std::stringstream original_ir;
+  original_ir << fusion;
+
+  // Test move before lowering
+  Fusion another_fusion = std::move(fusion);
```
Comment (jjsjann123) [inline test/cpp/jit/test_gpu.cpp:530]: nitpick: should we check move constructor does properly empty/clear the original?

**Review by jjsjann123 — APPROVED** [review]
LGTM.

File: torch/csrc/jit/codegen/cuda/ir_iostream.cpp (hunk 9 above)
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/ir_iostream.cpp]: Ok, renamed back to `vals`.

File: torch/csrc/jit/codegen/cuda/ir_base_nodes.cpp (hunk 11 above)
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/ir_base_nodes.cpp:213]: Oh, I see. Having a function which just does `this == other` doesn't seem to be worth it, so we may want to consider just replacing sameAs() directly with == where it makes sense.

I opened https://github.com/csarofeen/pytorch/issues/131 to track it.

**Comment by tlemo** [comment]
> Almost all objects in PyTorch even in the c++ backend follow this kind of pattern:
> ...
> Copies are only reference copies like they would be in ~pytorch~ python, not deep copies unless explicitly using clone().

Yes, that's because these are handle types ("pointers on steroids"). Value types >> handle types - what I mean is that you can easily implement the "handle" semantics on top of value types (ex. shared_ptr<Fusion>) but not the other way is not so easy.

I think that having Fusion as a fully-fledged value type may be a bit more complex but is beneficial (it allows uses that are not possible with simple handle types. Move semantics is one such example. Reasoning about thread safety is another. We can even do things like std::vector<Fusion> - as a cache backing store for example)

File: test/cpp/jit/test_gpu.cpp (hunk 12 above)
Comment (tlemo) [reply test/cpp/jit/test_gpu.cpp:530]: Good question! Yes, it's very tempted to do so, although generally the moved-away from objects are in a valid but unspecified state (for ex. in the standard library you can only call methods w/o preconditions, ex. clear() or assignment operators) - search for "preconditions" here: https://en.cppreference.com/w/cpp/utility/move

For this reason I originally decided against adding these checks, but at the same time it's easy to argue that test cases may afford to make extra assumptions. So I'm going ahead and adding the checks you suggested, thanks for bringing it up.

**Review by csarofeen — APPROVED** [review]

File: torch/csrc/jit/codegen/cuda/fusion.cpp [hunk 13]
```diff
@@ -57,6 +59,103 @@ std::unordered_set<Val*> InputsOf::output(Fusion* fusion, Val* output_) {
...
+    uses_.insert({val, std::move(val_uses)});
+  }
+
+  for (const auto& kv : other.values_map_) {
+    auto from_val = ir_cloner.clone(kv.first);
+    auto to_val = ir_cloner.clone(kv.second);
+    values_map_.insert({from_val, to_val});
+  }
+
+  inputs_ = ir_cloner.clone(other.inputs_);
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/fusion.cpp:139]: Mark to myself that I need to do the same for added containers in Fusion:
Three places to touch are `swap`, `clear` and copy constructor.
